### PR TITLE
NEW : Add status shipped in expedition

### PR DESCRIPTION
**Pull request by dolibarr95**
# New
Add a new status : SHIPPED

# Description
- Draft : prepare the expedition (choose the product to send from a propal or order)
- Validate : The expedition is in preparation (pick the products in warehouse, parcel, weight...) 
- Processed: The expedition is ready (to be send, chose the best carrier, tracking number...)
- Shipped: The expedition is shipped (the carrier pick the parcel)

 # Issue
#8838 

Hope everybody need this change #

### CODE REVIEW

**Comment by mwelters** [comment]
I think the most often used word for this status would be "SHIPPED"

**Comment by dolibarr95** [comment]
Hi
yes @mwelters you're right, it's done.

**Comment by dolibarr95** [comment]
Hi
If someone can help with transiflex (i really don't understand how to do).

# EN
- StatusSendingSendedShort : Shipped
- StatusSendingSended : Shipped
# FR
- StatusSendingSendedShort : Expédiée
- StatusSendingSended : Expédiée

File: htdocs/expedition/class/expedition.class.php [hunk 1]
```diff
@@ -1590,30 +1643,35 @@ function LibStatut($statut,$mode)
...
 			if ($statut==1) return $langs->trans($this->statuts[$statut]);
 			if ($statut==2) return $langs->trans($this->statuts[$statut]);
+			if ($statut==3) return $langs->trans($this->statuts[$statut]);
 		}
 		if ($mode==1)
 		{
 			if ($statut==0) return $langs->trans($this->statutshorts[$statut]);
 			if ($statut==1) return $langs->trans($this->statutshorts[$statut]);
 			if ($statut==2) return $langs->trans($this->statutshorts[$statut]);
+			if ($statut==23) return $langs->trans($this->statutshorts[$statut]);
```
Comment (frederic34) [inline htdocs/expedition/class/expedition.class.php]: 23 ?

**Comment by dolibarr95** [comment]
Hi
tks @frederic34  it's corrected now.

File: htdocs/expedition/class/expedition.class.php [hunk 2]
```diff
@@ -569,6 +572,56 @@ function fetch($id, $ref='', $ref_ext='', $ref_int='')
...
+
+		$resql=$this->db->query($sql);
+		if ($resql)
+		{
+			$this->statut=3;
+			
+			if (! $error)
+			{
+    				// Call trigger
+    				$result=$this->call_trigger('SHIPPING_SHIPPED',$user);
```
Comment (frederic34) [inline htdocs/expedition/class/expedition.class.php]: SHIPPING_SHIPPED, perhaps it's better to use SHIPPING_MODIFY wich is standard trigger

**Comment by dolibarr95** [comment]
@frederic34 tks for correction, It's done

**Comment by dolibarr95** [comment]
@eldy @frederic34 can you please give help for this conflict : 
**This branch has conflicts that must be resolved**
Where should i add this sql migration ?
```sql
ALTER TABLE llx_expedition ADD fk_user_shipped INT(11) DEFAULT NULL;
ALTER TABLE llx_expedition ADD date_shipped datetime DEFAULT NULL;
```

**Comment by frederic34** [comment]
try to add LF at end of file

**Comment by dolibarr95** [comment]
Hi
When I edit the doc there is already a line break (532 lines). But i will add an other one to test.

**Comment by frederic34** [comment]
do you know how to merge upstream/develop into your branch ? Doing that allow you to resolve conflict on your machine

**Comment by dolibarr95** [comment]
Hi
...sorry no i really don't know

**Comment by frederic34** [comment]
send me an email, i will explain

**Comment by dolibarr95** [comment]
tks a lot it's done.
I can give you write acces to my fork if you want.

**Comment by dolibarr95** [comment]
@frederic34  Tks a lot for your precious help.

**Comment by dolibarr95** [comment]
Hi
@eldy can you please check this PR please?
Tks

**Comment by dolibarr95** [comment]
@eldy Can you please help?
I don't understand why there is a conflict, i've resolved it but it's come back...
Tks for your help.

**Comment by dolibarr95** [comment]
Hi ! @eldy
Hope everybody enjoy holidays :)
Can an admin help me, tag or give opinion about this PR please ?
Have a nice day

Tks

**Comment by dolibarr95** [comment]
Conflicts solved 😀

**Comment by dolibarr95** [comment]
Hi
@atm-maxime can you please help ?

**Comment by ksar-ksar** [comment]
Hi,

When this PR will be merged ?
It is a must have feature I think.

**Comment by dolibarr95** [comment]
Hi,
Tks for your feed back, i wait for an admin opinion.
Hope soon

**Comment by ksar-ksar** [comment]
https://www.dolibarr.fr/forum/3-installation/62876-workflow-expedition

**Comment by dolibarr95** [comment]
@ksar-ksar up for this issue !
:)

**Comment by dolibarr95** [comment]
Hi everybody!

Little up for this Pr...
Is there an admin who can check or give advice/opinion/label/etc for this modifications ?

Have a nice monday

**Comment by dolibarr95** [comment]
Hi!
@FHenry can you please help ?

Have a nice day ;)

**Comment by ghost** [comment]
Seems to be nice. My guess is why for this status is need on expedition as it exists on order

**Comment by dolibarr95** [comment]
Hi!
_Tks for reading and for your answer._

Because we need to know the status of an expedition only (Draft, Validated, Done, Processed) because for an order we can send partial order and also the store keeper want's to make the difference between done and proced.
More informations, examples and explanations here #8838 
tks

**Comment by eldy** [comment]
A similar feature seems suggested to in #9660

**Comment by dolibarr95** [comment]
Hi
Tks for answer.
It's not the same #9660 is just about turn back in "draft" mode. What I suggest in this PR is to add a new status (4 status instead of 3 actually).
Tks for reading.

**Comment by eldy** [comment]
To analyze also with #9660 and #10174 and #8838

**Comment by eldy** [comment]
There is several PR to enhance this and it is not clear what will be the behaviour after (except that all of them suggest a different way of working). So to help to choose which one to retain, i updated wiki doc to describe current worklow:
https://wiki.dolibarr.org/index.php/Module_Shipments#Usage

Can you cut and paste this workflow text graph and update it to describe what will be workflow after the change (keeping the X and the number of status in parenthesis)

**Comment by eldy** [comment]
I think we should enhance the process with something like that:
Draft ---X---> Validated ---> Closed (X = stock change, Can edit qty and lot in Draft step)
must become
Draft ----> Validated ---X---> Shipped (X = stock change, Can edit qty and lot in status Draft + Validated)

**Comment by fbosman** [comment]
I would like to add a status: ready for shipment; the stock has been changed, but the products are still in the house.


I think we should enhance the process with something like that:
Draft ---X---> Validated ---> Closed (X = stock change, Can edit qty and lot in Draft step)
must become
Draft ----> Validated ---X---> ready for shipment ---> Shipped (X = stock change, Can edit qty and lot in status Draft + Validated)

**Comment by eldy** [comment]
It's strange to have X (the stock change) done before a status called "ready for shipment". ready for shipment means it is somewhere in a warehouse (may be a transitional warehouse) but not yet shipped. And a stock change reflect the fact that a product is no more in the hand of company (responsability of owning the product has been transferred to someone else).  But if it is not yet shipped, it is still somewhere. Imagine you received orders to order all you stock. When all your stocks will be "ready", if it is removed from physical stock ,the virtual stock will be (0 - n) (n all open orders), so the virtual stock shows we miss a lot of product and that we must replenish. But's its wrong. Virtual stock should be 0 since we have all quantity n in stock for n ordered so we have enough to send all orders done with no need of replenishment. So it will make very difficult to work this way.

Other software solve the need with 2 different solution (depending on software):
A) Some call this the status "Reserved" (it is still the property of the company, still in stock but not yet inside a transporter truck) but not available for others. And then they report 2 types of qty when getting availability of product (total for all or total for non reserved only). Often a gaz engine, because when doing an inventory, there is no way to know what are reserved or not.
B) Some other software or processes use a different solution by adding a type "common" or "transit" for warehouse. And they ask the name of the transit warehouse (with a possible default value if it is always the same "waiting area") when changing the status to something similar to "Ready for shipment", and they record a stock movement from common warehouses to transit warehouses. Then when they report stock available, they count only if warehouse is an available warehouse and exclude transitional warehouses.

It means, enhance the process with something like that:
Draft ---X---> Validated ---> Closed (X = stock change, Can edit qty and lot in Draft step)
must become
Draft ----> Validated ---Y---> ready for shipment -- X ---> Closed/Shipped (X = stock change, Y = stock move to a transit warehouse, We can edit qty and lot in status Draft + Validated).  
Adding this "step ---Y---> ready for shipment" could become easily an option (with or without the Y, we can decide later) is for me a second step. To make evolution of this second step easier to do, we must first move the X after the validated step, to get:
Draft ----> Validated ---X----> Closed/shipped (X = stock change, Can edit qty and lot in Draft+Validated step)

It will be easier after to insert a new status.

So, we must first have a PR that move the X after before think to insert a new step.

**Comment by BebZ** [comment]
Ahaha, Eldy we were reading the same subjects at the same time ! 
So much debate arround this subject ! 
Your POV is well described here, but I think everyone wants to modify a shipment BEFORE it is shipped (internal errors, customer wishes between the picking moment and the leave of the goods from the warehouse).

If there is something wrong after a shipment was done (so stock movement was recorded), and we need to send more, we should not go back to draft and resent the shipment. This has a lot of side effect.

Another solution to adresse that is to add option "Decrease real stocks on shipping classification to Closed / Shipped". When this option is on, you must be able to enter the batches before going to closed/shipped ?
Make it possible to modify the number shipped and lot when the shipment is in "Validated" status.

we must forecast a lot of use cases, but it make too much debates ! 
On the supplier reception side, it was less debated 
and it is much easier to make a reception and being able to put it back to draft, with stock reverts, correction, etc...

**Comment by eldy** [comment]
Current workflow of shipment is this one:
Draft --X--> Validated ---Y---> Shipped (We can choose to have stock change into X or Y step)

The Y has been implemented in v17 or before. You can now choose X or Y.

With v18, the default mode is set to Y instead of X.

What we need now is to be able edit qty and lot in status Draft + Validated so i close this PR (it seems some are already open for that so i close this PR).

**Pull request closed without merging**